### ISSUE TITLE
Music/sound stuttering or cut off (Retroarch)

### ISSUE THREAD

**Issue by KamiLordus** [flyinghead/flycast#1062]
Transitioning between locations, opening inventory, or maps noticeably cuts/stutter the music/sound.
I've tried different core options but nothing helps. I tested the game in CHD and GDI format.
I'm using RA 1.15 with the latest version of the Flycast core.
OS Win 11

**Comment by HyperSonic1999** [comment]
Also had this issue yesterday and just found a workaround. Disabling "Threaded Rendering" should stop it from cutting out when changing screens, only downside is that screen transitions take 2 or less seconds to load instead of just 1 second with TR on.

**Comment by KamiLordus** [comment]
@HyperSonic1999 thanks a lot. After disabling threaded rendering the game works normally.

**Comment by HyperSonic1999** [comment]
@KamiLordus No prob, hope you enjoy.

**Comment by flyinghead** [comment]
reopening since this doesn't look right

**Comment by sappergr** [comment]
Using the flatpak version of Flycast (not sure if relevant) Videos seems to freeze if not using VGA cable method.

**Comment by mfigueirido** [comment]
Exact same behaviour on Alone In The Dark - The New Nightmare. There's lots of audio stuttering specially noticable when changing screens. Fixed by disabling threaded rendering but then screen load times go from almost instant to taking around a second.

**Comment by tmmmmm** [comment]
Installing a bios-file fixed the video/audio-stuttering for me.
(on Steam Deck, running Flycast v2.3 standalone)

**Comment by TessArdent** [comment]
Although disabling thread rendering does fix it, "Detect Frame Rate Changes" is a much better solution. The audio transition is still not smooth, but I think that's just a symptom of the game loading the next angle much quicker.
Set `Core Options > Video > Detect Frame Rate Changes` to OFF.
I also added dc_boot and dc_flash to my bios. I'm not sure if that's necessary, but it did add the dreamcast opener when I launch RE2.

**Comment by mfigueirido** [comment]
Made some more testing regarding this with refreshed Retroarch and Flycast core. Used a real bios file and "Detect Frame Rate Changes" was off. On my side there's simply no way to get clean audio when using the RA core. With "Threaded Rendering" enabled the audio cuts during any kind of transition, and if you disable it the sound stuttering kicks in, super easy to spot during RE2 intro FMV, but happens everywhere.

This happens as well in Alone in the Dark, Sega Rally 2 and probably many other games:
https://github.com/flyinghead/flycast/issues/1211

**Comment by mfigueirido** [comment]
Managed to get it working right. This is what I did:

1. Disable "Threaded Rendering".
2. In Retroarch's menu go to Settings->Video->Synchronization and set "Max Swapchain Images" to 4. If you don't do this some sporadic stuttering will remain.
3. Ensure you don't have the SH4 CPU overclocked, or the overclock is not big at least. In my tests Resident Evil 2 could be overclocked up to 300Mhz with no sound stuttering, while Sega Rally 2 only accepts up to 250 Mhz.

This way the game will play perfectly except that when switching camera angles there will be some slowdown that "Threaded Rendering" fixes.

**Comment by KamiLordus** [comment]
After longer tests I noticed that the problem is not limited to RE 2, but many other games.
As @mfigueirido noticed, disabling "Threaded Rendering" helps, but also significantly slows down many more demanding games, which cannot maintain a constant 60 fps without this option.
From what I have noticed, "Threaded Rendering" does not have to be disabled, because enabling the "Full Frame Buffer" option also helps to fix the sound problem. Unfortunately, this is not a perfect solution, because when I enable it on the GL Core graphics driver, the number of fps drops below 55 and the sound still stutters. The problem does not occur on the Vulkan driver (which is more efficient). However, from my observations, not every game likes the "Full Frame Buffer" option enabled and can cause graphical errors (for example, the character select screen in Dead or Alive 2).

**Comment by flyinghead** [comment]
This is due to the flycast core only synchronizing on the rendering, and not on audio (as standalone does). When the game isn't rendering (loading, transitions, etc.) then audio is lost.
I'm not responsible for this change.

**Comment by KamiLordus** [comment]
@flyinghead, getting back to the point, do you mean it's an RA issue? If so, could you tell me more about what the problem is?

**Comment by flyinghead** [comment]
I believe the issue is due to the flycast core only synchronizing on vsync. Standalone Flycast uses both vsync (if enabled) and audio output to synchronize the emulator with real time.
So when a game isn't rendering, or rendering at a very slow rate, which is typical of transitions and loading, the core doesn't have anything to synchronize on and runs in fast forward.
I'm not 100% sure that all issues mentioned in this thread are due to this problem but I bet a number of them are.
The best way to verify is to compare RA and standalone with similar settings on those games and see if the problem also happens with standalone.

**Comment by KamiLordus** [comment]
Thanks for the reply and detailed explanation.
Although, in my case, I wasn't able to fix this problem using any of the v-sync settings in RA.
I'm currently using a 240Hz monitor with g-sync, and the problem is the same whether g-sync is on or off. The same applies to regular v-sync on/off; it doesn't matter.

The only solution here is to disable threaded rendering, which fixes the problem in every game that has this problem (and that I can quickly recall).

Resident Evil 2
Samurai Shodown 6
Fist of the North Star
Crazy Taxi
Crazy Taxi 2
And probably many others.

Unfortunately, this isn't a perfect solution. It's likely that after disabling threaded rendering, not every system will be able to achieve a satisfactory FPS.
I'll try to test the standalone version of Flycasa soon; I'm curious if the problem persists there as well.

Edit:
Okay, I just tested the standalone emulator with the same settings as I have on the Flycast core in RA.
I didn't notice any issues with multi-threat emulation enabled.
However, in the case of RE 2, transitions are slower, practically the same as on the Flycast core in RA when threaded rendering is disabled.
Perhaps it's a bug in RA and the frontend isn't handling threaded rendering synchronization very well.

**Comment by joepogo** [comment]
@flyinghead @KamiLordus 

I came here to comment because I'm having weird issues similar to this on my steamOS mini PC and my laptop both being powerful enough to run DC games. Id figured I'd ask before opening this as an issue beforehand.

If you load up Chu chu rocket for instance it's extremely noticeable when transitioning from menus before you start a game, the sound and fps drops causing stuttery choppy sound. Once in game, you're fine but it's enough to be noticable and probably shouldn't be happening.

I'm not using run ahead, threaded rendering or any other settings toggled like that.


Just to clarify as well I've tested both the newest GitHub actions build of retroarch flycast from flying head's GitHub AND the current version of retroarch flycast available through the nightly buildbot all with the same results. 

Obviously I tested this against the standalone version of flycast and while the fps dips between menus like normal, the sound is normal.

**Comment by JoeMatt** [comment]
There is a technical solution here, my fork I'm using for iFly decouples audio from stalling on slow frames, in native frontend and the retoarch api

Unfortunately I have to keep the code local for now, since last time I pused flycast for iOS updates another app pushed them publically first, before I was really done, and forced my hand to support and push my own copy

Once iFly is out, I want to see what to upstream to flycast, things that aren't iOS/tvOS specific, of which there are many many fixes

Will take a good LLM to cross review and break them apart but I'm confident issues like this one should port over easily, but it will be a big project to do it in a linear and smart way that doesn't overload flyinghead with complex PRs

**Comment by joepogo** [comment]
This is great @JoeMatt ! Looking forward to seeing this in action. Thank for responding here, makes me feel like I'm not totally crazy. 😆 

It made me think my external SSD was going wonky but trying out internal storage yields the same results.

**Comment by JoeMatt** [comment]
> This is great [@JoeMatt](https://github.com/JoeMatt) ! Looking forward to seeing this in action. Thank for responding here, makes me feel like I'm not totally crazy. 😆
> 
> It made me think my external SSD was going wonky but trying out internal storage yields the same results.

Oh yeah, def not crazy.

I think it's this but not only this change,

https://github.com/JoeMatt/flycast/commit/9da1150a3481a4761a66b3c0f75c02aef3f5ea2b

and probably this,
https://github.com/JoeMatt/flycast/commit/9d245ec4f2086eafeba92a6427ee583aee6f4b0b#diff-d7508962b121da3966ec72416b6833fef25e9805da3f6a34333476fb8c9ecba2

but a platform agnostic version

**Comment by joepogo** [comment]
Those two links return a 404 error for some reason. 

I know at the moment mainline libretro flycast won't have this fix but if you integrate this into flycast anytime soon please let me know, can never have enough dll cores and I'd love to try it. 🤝👍

It's been a problem that's bothered me more than it should for a while now. 😀

**Comment by JoeMatt** [comment]
Oh yeah I think that branch is private now for the moment, I'll open it up once I ship the final version, weeks not months

**Comment by joepogo** [comment]
Ah, gotcha! Nonetheless it's awesome to see this in the works. Thank you for responding on here, I'm definitely keen to see it. 

Flyingheads work is too notch it's just the issue has plagued me a long time. 🤏😀

**Comment by KamiLordus** [comment]
@joepogo Yes, it's the same problem, and I mentioned it in my previous post.
It only occurs in RA, and the standalone Flycast version is free of this issue.
Unfortunately, I've reported this to the Libretro admins multiple times, and nothing has changed. Only disabling multithreats helps, but some games will be too slow/unplayable.

**Comment by joepogo** [comment]
i actually think this can be rectified by adding a pre-cache option for chd, as a test i added this in a test core and it alleviated the issues, its only about a 10 second wait to load but after that its gravy! still testing though but all the stuttering was gone after that.
